### ISSUE TITLE
EVP_CIPHER's initialization with variable-length keys and IVs is dangerous and not memory-safe

### ISSUE THREAD

**Issue by davidben** [openssl/openssl#22629]
This is a follow-up to CVE-2023-5363. I think the root cause here is actually a design flaw in how OpenSSL modeled variable-length IVs. With the new `OSSL_PARAM` system, it has gotten quite dangerous.

OpenSSL's original `EVP_CIPHER` initialization APIs looked like this:
```
int EVP_EncryptInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                    const unsigned char *key, const unsigned char *iv);
```
As in #22628, this requires the caller to know the sizes for `cipher` and size the key and IV correctly. Most algorithms have fixed-width keys and fixed-width IVs, and sizing is less complex than Update/Final, so this wasn't _horrible_, just a bit of a sharp edge.

However, some ciphers have variable-length keys and IVs. RC4 has a variable-length key, and AES-GCM was incorrectly specified with a variable-length IV. This API design cannot handle such algorithms. OpenSSL ended up handling this by adding `EVP_CIPHER_CTX_set_key_length` and `EVP_CTRL_AEAD_SET_IVLEN`. This means initializing an `EVP_CIPHER_CTX`, in the general case, becomes:

1. `EVP_EncryptInit` to set the cipher, but leave key and IV null and unset
2. `EVP_CIPHER_CTX_set_key_length` and `EVP_CTRL_AEAD_SET_IVLEN`
3. `EVP_EncryptInit` again. This time leave cipher null, because you've already set it, and set the key and IV.

This is fairly tedious, but works. Since then, OpenSSL 3.x has converted key and IV length to `OSSL_PARAM`s, with an API of:

```
int EVP_EncryptInit_ex2(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                        const unsigned char *key,
                        const unsigned char *iv,
                        const OSSL_PARAM params[]);
```

On the one hand, this is single-shot and let's you combine 1-3 (but not really, more on this later...). But, in doing so, it makes the API incredibly dangerous! The sizes of `key` and `iv` are now buried inside `params`, behind a bunch of strings! Now, to answer whether this call is correct, you need to look for "keylen" and "ivlen" inside `params`.

Worse, because providers typically silently ignore unused parameters (see https://github.com/openssl/openssl/issues/16912#issuecomment-951172754), you may have set the right "ivlen" and then the provider ignored it anyway!

This is confusing enough that OpenSSL itself forgot all this, leading to CVE-2023-5363. `OSSL_PARAM` looks enough like a well-typed, dynamic system that it'd be very tempting for bindings libraries to simply expose them directly in their public APIs. But the `EVP_EncryptInit_ex2` design means that doing so is not memory-safe, or at least very difficult to do safely. (It might work for a bindings library to go back to the multi-step model and leave key/iv unset until they can query the lengths that actually ended up? Not sure. But this is very non-obvious.)

How could this have been done better? Well, once it's been decided keys and IVs are variable-length, simply use pointer/length pairs like any other C API:
```
int EVP_EncryptInit_ex1000(
    EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
    const unsigned char *key, size_t key_len,
    const unsigned char *iv, size_t iv_len,
    ....);                 
```

Or if you prefer `OSSL_PARAM`, make "key" and "iv" byte string properties. Just don't do half-and-half. Either consistently use a real parameter, or consistently use an `OSSL_PARAM`. This is enough of an issue that new APIs are probably worth making.

## A second API design issue

Now, if you're going to make new APIs, it may be worth revisiting another design flaw with initialization. (We don't do `OSSL_PARAM` for performance, but the multi-step setup is quite tedious.) `EVP_CIPHER` fails to distinguish per-key state and per-operation state.

It is very common for applications to encrypt multiple messages with a single key, and OpenSSL's API is poorly suited for this. Most algorithms have some per-key computations (e.g. the AES key schedule) and some per-operation computations. Ideal behavior would be to do the per-key computations only once per key, and hold on to that. That means the general API pattern is actually:

1. Setup the key:
   1. `EVP_EncryptInit(cipher, rest null)`
   2. `EVP_CIPHER_CTX_set_key_length`
   3. `EVP_EncryptInit(key, rest null)`
2. For each encryption operation:
   1. `EVP_CTRL_AEAD_SET_IVLEN`
   2. `EVP_EncryptInit(iv, rest null)`
   3. `EVP_EncryptUpdate`, `EVP_EncryptFinal`

Except even this does not work for all applications! The encryption procedure is mutating. But we don't want to mutate the saved key state. That may not even be thread-safe. So actually the procedure is:

1. Setup the key:
   1. `EVP_EncryptInit(cipher, rest null)`
   2. `EVP_CIPHER_CTX_set_key_length`
   3. `EVP_EncryptInit(key, rest null)`
2. For each encryption operation:
   1. Copy the `EVP_CIPHER_CTX`
   2. `EVP_CTRL_AEAD_SET_IVLEN`
   3. `EVP_EncryptInit(iv, rest null)`
   4. `EVP_EncryptUpdate`, `EVP_EncryptFinal`

Not only that, the per-key `EVP_CIPHER_CTX` still maintains the per-operation state, even though it doesn't need it. So we waste memory. Also because this is needed to use `EVP_CIPHER_CTX` effectively, `EVP_CIPHER` is forced to be initializable piecemeal, which lets a very complex state machine and weird behaviors like #22070.

## Very (very) rough sketch

A better version of `EVP_CIPHER`, that better models the actual requirements around initialization, might look something like this:

First, there are two types, `EVP_CIPHER_KEY` and `EVP_CIPHER_CTX`. The former is a key, and the latter is a streaming encrypt/decrypt operation.

To create an `EVP_CIPHER_KEY`, there are functions like:
```
EVP_CIPHER_KEY *EVP_CIPHER_new_enc_key(const EVP_CIPHER *cipher, const uint8_t *key, size_t key_len);
EVP_CIPHER_KEY *EVP_CIPHER_new_dec_key(const EVP_CIPHER *cipher, const uint8_t *key, size_t key_len);
```
This is just a sketch. Maybe insert `OSSL_PARAM`s if you prefer. Maybe merge them into one operation with an `enc` boolean. The point is this is just the key and it's immutable on creation. And also rather than setting the key length separately, it always comes with the key pointer.

To do an operation, there are functions like:
```
EVP_CIPHER_CTX *EVP_CIPHER_KEY_new_ctx(const EVP_CIPHER_KEY *key, const uint8_t *iv, size_t iv_len);
```
Again, just a sketch. Maybe `OSSL_PARAM`s if you prefer, etc. The point is this is a separate operation from setting up the `EVP_CIPHER_KEY`, and that the IV length always comes with the IV pointer.

Given an `EVP_CIPHER_KEY` type, you could even make friendlier single-shot encrypt/decrypt APIs, for callers that don't need anything streaming.

**Comment by mattcaswell** [comment]
This sounds very like the EVP_SKEY proposal in #19754, and on the project board as https://github.com/openssl/project/issues/104, i.e. this is already something that is on the radar as something to implement.

**Comment by beldmit** [comment]
See also https://github.com/openssl/openssl/pull/21573

**Comment by davidben** [comment]
@beldmit That design document seems not to address the memory-safety flaws in OpenSSL's API. The proposed `EVP_CipherInit_skey` continues the pattern of a pointer-only IV parameter.

**Comment by davidben** [comment]
I.e. the `EVP_SKEY`/`EVP_CIPHER_KEY`/whatever proposal is not the point of this bug. That is a secondary issue that, if you're making a new API anyway, you should take the time to rethink it. But the actual problem, and the unaddressed root cause of CVE-2023-5363, is OpenSSL APIs losing track of the length associated with each pointer.

**Comment by beldmit** [comment]
@davidben yes, being drafted several months ago, this document needs a review and I will update it according to your requirements

**Comment by beldmit** [comment]
> @beldmit That design document seems not to address the memory-safety flaws in OpenSSL's API. The proposed `EVP_CipherInit_skey` continues the pattern of a pointer-only IV parameter.

Thanks, fixed

**Comment by davidben** [comment]
Might need some discussion around how that interacts with the "ivlen" OSSL_PARAM, since that function still takes a parameter list.

**Comment by beldmit** [comment]
Yes, for the uniformity it may be worth passing IV through the parameters

**Comment by davidben** [comment]
That may have some perf implications since IVs are common. But an "iv" parameter would still need to answer what happens with "ivlen" in OSSL_PARAM. It should be the case that setting "ivlen" in params does not result in a memory error, whether it's an "iv" OSSL_PARAM or a real (i.e. more efficient) parameter.

**Comment by beldmit** [comment]
@davidben I'm not sure parameters parsing will cause performance implications in real-world scenarios.

**Comment by t8m** [comment]
IV should be kept as an explicit function argument. I do not think we should change it to be an OSSL_PARAM. The "ivlen" OSSL_PARAM could be made obsolete with the new APIs and for backwards compatibility we could just call set_ctx_params with "ivlen" param if the ivlen in the arguments does not match the default (or previous in case of reinitialization) ivlen.

**Comment by t8m** [comment]
IMO this is resolved in the new API calls using EVP_SKEY.

**Comment by beldmit** [comment]
Yes, and we will have to deprecate the old API at some moment...
